README.md
# Gatsby Remark oEmbed Plugin

_Drop a link to oEmbed content (Twitter, Instagram, YouTube, Vimeo, SoundCloud, CodePen etc.) and see it transform into proper embed html._

&nbsp;  
🚨 gatsby-plugin-mdx support is lacking due to problems with Gatsby Cache in sub-plugins.

&nbsp;

## A message or two from Queen Raae 👑

### 1-on-1 Emergency Gatsby Call

Are you stuck on a reef in the sharky waters around the Gatsby islands? Check out [1-on-1 Emergency Gatsby Call](https://queen.raae.codes/gatsby-emergency/) with Queen Raae to get friendly advice you can put into action immediately from a seasoned Gatsby developer.

### Stay updated and get the most out of Gatsby

Learn how to get the most out of Gatsby and **stay updated** on the plugin by [subscribing](https://queen.raae.codes/emails/?utm_source=readme&utm_campaign=remark-oembed) to daily emails from Queen Raae and Cap'n Ola.

&nbsp;

## A note on oEmbed

> oEmbed is a format for allowing an embedded representation of a URL on third party sites. The simple API allows a website to display embedded content (such as photos or videos) when a user posts a link to that resource, without having to parse the resource directly.

Learn more by checking out [oembed.com](https://oembed.com).

Links dropped into your markdown content is checked against the oEmbed provider list from [oembed.com](https://oembed.com/#section7). If there is a match a call is made to the provider's oEmbed endpoint and the link is swapped for the embed html.

The provider list is downloaded on every build so that oEmbed support stays up to date.

So far these providers are confirmed to be working: CodePen, Flickr, Instagram, Reddit, Twitter, Vimeo, YouTube, SoundCloud.

Twitter, Flickr, Instagram and Reddit requires external javascript to be added to every page. So make sure to exclude the ones you do not need.

> **Warning:** The Instagram oEmbed API requires an access token. You'll need a Facebook Developer account, a Facebook App, and an App Access Token. [More information here](https://developers.facebook.com/docs/instagram/oembed), or take a look at this [tutorial for setting it all up in Gatsby with Environment Variables](https://phil.tech/2020/gatsby-instagram-oembed-failures/).

> **Update:** Twitch removed oEmbed support in 2020 when deprecating their v5 API. It seems they are not planning to support oEmbed again. Unfortunately this means Twitch urls are no longer being transformed. Let them know how you feel about this [on their forum](https://discuss.dev.twitch.tv/t/oembed-deprecation/24424/2).

&nbsp;

## How to install

`npm install @raae/gatsby-remark-oembed`

or

`yarn add @raae/gatsby-remark-oembed`

## Requirements

- Node version 8 and up.
- Gatsby version >=2.0.88
- Gatsby Transformer Remark Plugin >=2.0.0

&nbsp;

## How to use

### Configuration

```js
// In your gatsby-config.js
plugins: [
  {
    resolve: `gatsby-transformer-remark`,
    options: {
      plugins: [
        {
          resolve: `@raae/gatsby-remark-oembed`,
          options: {
            // usePrefix defaults to false
            // usePrefix: true is the same as ["oembed"]
            usePrefix: ["oembed", "video"],
            providers: {
              // Important to exclude providers
              // that adds js to the page.
              // If you do not need them.
              exclude: ["Reddit"],
            },
          },
        },
      ],
    },
  },
];
```

#### Settings per provider

Many oEmbed providers offer additional options for configure the display of the embed.

For example, for Twitter see [Embedded Tweet parameter reference](https://developer.twitter.com/en/docs/twitter-for-websites/embedded-tweets/guides/embedded-tweet-parameter-reference), which describes the additional oEmbed parameters you might want to change for the embed.

```js
// …
{
  resolve: `@raae/gatsby-remark-oembed`,
  options: {
    usePrefix: true,
    providers: {
      include: [
        'Twitter',
        'Instagram',
      ],
      settings: {
        // Ex. Show all Twitter embeds with the dark theme, and disables ad tracking
        Twitter: {
          theme: 'dark',
          dnt: true
        },
        // Ex. Hide all Instagram comments by default, and
        // add a facebook access token
        Instagram: {
           hidecaption: true,
           access_token: 'a-facebook-access-token'
        },
      },
    },
  },
}
```

### Content

#### With setting `usePrefix: true`

```md
// In your markdown file

Check it out! I can use oembed links in my markdown.

`oembed: https://twitter.com/raae/status/1045394833001652225`

Its pretty cool :D

`oembed: https://www.instagram.com/p/Bof9WhgBmY2/`
```

Links must be surrounded by empty lines.

#### With setting `usePrefix: array of prefixes`

If you would like to use a prefix other than "oembed" or multiple prefixes you can set `usePrefix` to an array of custom prefixes. This can be beneficial when converting from other embed plugins.

`usePrefix: ["embed", "video", "oembed"]`

```md
// In your markdown file

Check it out! I can use the prefix "oembed:"

`oembed: https://twitter.com/raae/status/1045394833001652225`

Or I can use the prefix "embed:" if I like ;)

`embed: https://www.instagram.com/p/Bof9WhgBmY2/`

I can also use "video:" like I did before with `gatsby-remark-video`.

`video: https://vimeo.com/42672205`
```

`usePrefix: true` in the section above is the same as `usePrefix: ["oembed"]`.

#### With setting `usePrefix: false`

```md
// In your markdown file

Check it out! I can use oembed links in my markdown.

https://twitter.com/raae/status/1045394833001652225

Its pretty cool :D

https://www.instagram.com/p/Bof9WhgBmY2/
```

Links must be surrounded by empty lines.

## Options

| Name                 | Type                        | Description                                                 |
| -------------------- | --------------------------- | ----------------------------------------------------------- |
| `usePrefix`          | Boolean / Array of prefixes | See above section on content                                |
| `providers.include`  | Array of provider keys      | Only links from providers on this list will be transformed. |
| `providers.exclude`  | Array of provider keys      | Links from providers on this list will not be transformed.  |
| `providers.settings` | Object of provider settings | Optional configuration unique to each provider.             |

&nbsp;

## How to contribute

If you would like to contribute with code, check out `CONTRIBUTING.md`.

### Questions, Feedback and Suggestions

Do not be shy!

If you have any questions, feedback or suggestions head on over to [discussions](https://github.com/queen-raae/gatsby-plugin-let-it-snow/discussions).

### Bugs

If you find a bug please open an [issue](https://github.com/raae/gatsby-plugin-let-it-snow/issues) and/or create a pull request to fix it.


CHANGELOG.md
## [0.3.3](https://github.com/queen-raae/gatsby-remark-oembed/compare/v0.3.2...v0.3.3) (2022-12-26)

## [0.3.2](https://github.com/queen-raae/gatsby-remark-oembed/compare/v0.3.1...v0.3.2) (2022-07-20)


### Bug Fixes

* transform paragraph parent node for standalone link nodes ([#170](https://github.com/queen-raae/gatsby-remark-oembed/issues/170)) ([193547e](https://github.com/queen-raae/gatsby-remark-oembed/commit/193547e81d4d90a78d612070aa9b69196ceda1cd)), closes [#102](https://github.com/queen-raae/gatsby-remark-oembed/issues/102)

## [0.3.1](https://github.com/queen-raae/gatsby-remark-oembed/compare/v0.3.0...v0.3.1) (2022-04-28)


### Bug Fixes

*  remove console.log ([3bd216b](https://github.com/queen-raae/gatsby-remark-oembed/commit/3bd216b9a88d37d12d81ce26bf5cf43dde75bb9b))

# [0.3.0](https://github.com/queen-raae/gatsby-remark-oembed/compare/v0.2.1...v0.3.0) (2022-04-28)


### Features

* find only standalone text links in root paragraphs ([#161](https://github.com/queen-raae/gatsby-remark-oembed/issues/161)) ([dd60700](https://github.com/queen-raae/gatsby-remark-oembed/commit/dd607001b847fad7740a9940c33f7eaa78351668)), closes [#160](https://github.com/queen-raae/gatsby-remark-oembed/issues/160) [#162](https://github.com/queen-raae/gatsby-remark-oembed/issues/162)

## [0.2.1](https://github.com/queen-raae/gatsby-remark-oembed/compare/v0.2.0...v0.2.1) (2022-04-28)

# [0.2.0](https://github.com/queen-raae/gatsby-remark-oembed/compare/v0.1.1...v0.2.0) (2022-04-28)


### Features

* new beginnings ([6edd31a](https://github.com/queen-raae/gatsby-remark-oembed/commit/6edd31a686882d1dc72adbf2ca534b559b8e75f8))


helper.test.data.js
exports.MARKDOWN_AST = {
  type: "root",
  children: [
    {
      type: "paragraph",
      children: [
        {
          type: "text",
          value: "Take a look at a day in my life as a web developer.",
          position: {
            start: { line: 2, column: 1, offset: 1 },
            end: { line: 2, column: 52, offset: 52 },
            indent: []
          }
        }
      ],
      position: {
        start: { line: 2, column: 1, offset: 1 },
        end: { line: 2, column: 52, offset: 52 },
        indent: []
      }
    },
    {
      type: "paragraph",
      children: [
        {
          type: "link",
          title: null,
          url: "https://www.instagram.com/p/Bof9WhgBmY2",
          children: [
            {
              type: "text",
              value: "https://www.instagram.com/p/Bof9WhgBmY2",
              position: {
                start: { line: 4, column: 1, offset: 54 },
                end: { line: 4, column: 40, offset: 93 },
                indent: []
              }
            }
          ],
          position: {
            start: { line: 4, column: 1, offset: 54 },
            end: { line: 4, column: 40, offset: 93 },
            indent: []
          }
        }
      ],
      position: {
        start: { line: 4, column: 1, offset: 54 },
        end: { line: 4, column: 40, offset: 93 },
        indent: []
      }
    },
    {
      type: "paragraph",
      children: [
        {
          type: "text",
          value: "Follow me for more code fun at ",
          position: {
            start: { line: 6, column: 1, offset: 95 },
            end: { line: 6, column: 32, offset: 126 },
            indent: []
          }
        },
        {
          type: "link",
          title: null,
          url: "https://raae.codes",
          children: [
            {
              type: "text",
              value: "raae.codes",
              position: {
                start: { line: 6, column: 33, offset: 127 },
                end: { line: 6, column: 43, offset: 137 },
                indent: []
              }
            }
          ],
          position: {
            start: { line: 6, column: 32, offset: 126 },
            end: { line: 6, column: 64, offset: 158 },
            indent: []
          }
        },
        {
          type: "text",
          value: ".",
          position: {
            start: { line: 6, column: 64, offset: 158 },
            end: { line: 6, column: 65, offset: 159 },
            indent: []
          }
        }
      ],
      position: {
        start: { line: 6, column: 1, offset: 95 },
        end: { line: 6, column: 65, offset: 159 },
        indent: []
      }
    },
    {
      type: "paragraph",
      children: [
        {
          type: "inlineCode",
          value: "oembed: https://twitter.com/raae/status/1045394833001652225",
          position: {
            start: { line: 8, column: 1, offset: 161 },
            end: { line: 8, column: 62, offset: 222 },
            indent: []
          }
        }
      ],
      position: {
        start: { line: 8, column: 1, offset: 161 },
        end: { line: 8, column: 62, offset: 222 },
        indent: []
      }
    },
    {
      type: "paragraph",
      children: [
        {
          type: "inlineCode",
          value: "oembed:https://www.instagram.com/p/Bof9WhgBmY2",
          position: {
            start: { line: 10, column: 1, offset: 224 },
            end: { line: 10, column: 49, offset: 272 },
            indent: []
          }
        }
      ],
      position: {
        start: { line: 10, column: 1, offset: 224 },
        end: { line: 10, column: 49, offset: 272 },
        indent: []
      }
    }
  ],
  position: {
    start: { line: 1, column: 1, offset: 0 },
    end: { line: 11, column: 1, offset: 273 }
  }
};

exports.PROVIDERS = [
  {
    provider_name: "Instagram",
    provider_url: "https://instagram.com",
    endpoints: [
      {
        schemes: [
          "http://instagram.com/p/*",
          "http://instagr.am/p/*",
          "http://www.instagram.com/p/*",
          "http://www.instagr.am/p/*",
          "https://instagram.com/p/*",
          "https://instagr.am/p/*",
          "https://www.instagram.com/p/*",
          "https://www.instagr.am/p/*"
        ],
        url: "https://api.instagram.com/oembed",
        formats: ["json"]
      }
    ]
  },
  {
    provider_name: "Kickstarter",
    provider_url: "http://www.kickstarter.com",
    endpoints: [
      {
        schemes: ["http://www.kickstarter.com/projects/*"],
        url: "http://www.kickstarter.com/services/oembed"
      }
    ]
  },
  {
    provider_name: "Twitter",
    provider_url: "http://www.twitter.com/",
    endpoints: [
      {
        schemes: [
          "https://twitter.com/*/status/*",
          "https://*.twitter.com/*/status/*"
        ],
        url: "https://publish.twitter.com/oembed"
      }
    ]
  }
];
